protocol/validation/validation.go

```diff
@@ -53,169 +55,208 @@ var (
 	errZeroTime              = errors.New("timerange has one or two bounds set to zero")
 )
 
-func checkValid(vs *validationState, e bc.Entry) error {
+func checkValid(vs *validationState, e bc.Entry) (err error) {
+	entryID := bc.EntryID(e)
+	var ok bool
+	if err, ok = vs.cache[entryID]; ok {
+		return err
+	}
+
+	defer func() {
+		vs.cache[entryID] = err
+	}()
+
 	switch e := e.(type) {
 	case *bc.TxHeader:
 		// This does only part of the work of validating a tx header. The
 		// block-related parts of tx validation are in ValidateBlock.
 		if e.MaxTimeMs > 0 {
 			if e.MaxTimeMs < e.MinTimeMs {
-				return errors.WithDetailf(errBadTimeRange, "min time %d, max time %d", e.MinTimeMs, e.MaxTimeMs)
+				err = errors.WithDetailf(errBadTimeRange, "min time %d, max time %d", e.MinTimeMs, e.MaxTimeMs)
+				return
```

Comment:
It would look nicer to leave these return statements as they were. Most of them don't need to change.